.github/workflows/docs.yml

```diff
@@ -3,41 +3,35 @@ on:
   push:
     branches: [main]
   pull_request:
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.head.label || github.head_ref || github.ref }}
+  cancel-in-progress: true
 jobs:
   docs:
     permissions:
       contents: write
-    runs-on: ubuntu-latest
+      id-token: write
+    runs-on: ubuntu-latest-4-cores
     env:
-      RUST_TOOLCHAIN: "nightly-2023-10-26"
       RUSTDOCFLAGS: "-D warnings"
+      RUSTC_WRAPPER: "sccache"
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v3
+        uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac #v4
```

Comment:
Out of curious: Why pointing to a hash directly?

Reply:
In the unlikely case that the actions/checkout repository introduces a bad update or let's say someone hacks it and introduces malicious code, our pipeline will run the update without questions if we do `v1` vs the code commit. It's a way of freezing the code we run from others.
Keep in mind some of the CI pipelines have access to our Google Cloud credentials so a malicious change can potentially obtain access too.
More info: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions